[KOGITO-9464] Update Operator Guides and Use Cases (#381)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/cloud/operator/referencing-resource-files.adoc

> = Referencing Resource Files in the Workflow

Reviewer: Is "resource files" a description people will understand?  Not sure about a better name though
```suggestion
= Referencing Additional Files from the Workflow
```

:compat-mode!:
// Metadata:
:description: Referencing external resource files in a workflow definition
:keywords: workflow, serverless, operator, kubernetes, minikube, configmap
// links
:kubernetes_create_configmap_url: https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-a-configmap
:kubernetes_configmap_url: https://kubernetes.io/docs/concepts/configuration/configmap/

This document describes how to reference resources files in the `SonataFlow` Custom Resource (CR).

> Most of the time, a workflow definition will required not only the flow definition, but also specification descriptors, schemas, subflows definitions, and etc.

Reviewer 2: ```suggestion
Most of the time, a workflow definition will required not only the flow definition, but also OpenAPI specification descriptors, schemas, sub-flow definitions, etc.
```

For example, given the following workflow:

.Example of a workflow referencing external resources
[source,yaml,subs="attributes+"]
----
apiVersion: sonataflow.org/v1alpha08
kind: SonataFlow
metadata:
  name: service
  annotations:
    sonataflow.org/description: Hello Service!
    sonataflow.org/version: 0.0.1
    sonataflow.org/profile: dev
spec:
  flow:
    start: Service
    dataInputSchema: specs/workflow-service-schema.json <1>
    functions:
    - name: isWinner
      operation: specs/workflow-service-openapi.json#isWinner <2>
      type: rest
    states:
    - name: Service
      type: operation
      actions:
      - name: CallService
        functionRef:
          refName: isWinner
    end: true
----

<1> The workflow defines an input schema
<2> The workflow requires an OpenAPI specification file to make a REST invocation

For this example, you have two options. You can either create two `ConfigMaps` to have a clear separation of concerns or only one with both files.

> From the operator perspective, it won't make any difference since both files will be available for the workflow application in runtime.

Reviewer 3: @ricardozanini `at runtime` rather than `in runtime`?

To make it simple, you can create only one `ConfigMap`. Given that the files are available in the current directory:

.Creating a ConfigMap from the current directory
[source,bash,subs="attributes+"]
----

> kubectl create configmap service-resources --from-file=$(pwd) -n <workflow-namespace>

Reviewer 2: $(pwd) might not be a good idea.
if user are running this command from a dir that have more files, that can lead the user to have issues.
maybe enforce to create a directory and add the files there would be nice.

Author: But that's the point. They created a new dir in the past command, added the files in the current dir and create the ConfigMap.

spec:
  resources: <1>
    configMaps:
      - configMap:
          name: service-resources <2>
        workflowPath: specs <3>
  flow:
    start: Service
    dataInputSchema: specs/workflow-service-schema.json
    functions:
    - name: isWinner
      operation: specs/workflow-service-openapi.json#isWinner
      type: rest
    states:
    - name: Service
      type: operation
      actions:
      - name: CallService
        functionRef:
          refName: isWinner
    end: true
----

<1> Introduced a new attribute `.spec.resources` where you can bind the `ConfigMap` to the `SonataFlow` CR
<2> The name of the `ConfigMap` in the same namespace
<3> The path where we want to reference these resources

Note that the `workflowPath` is `specs`. This is the path where you want to reference the files within the `ConfigMap` in the workflow definition.

[IMPORTANT]
====
Always create your `ConfigMaps` before the `SonataFlow` since not having the resources available during startup might break the deployment.
====

Any resource you have to map to the flow definition can be added to the `SonataFlow` CR using this procedure.

== Creating a Static Service Registry

> The `ConfigMap` containing workflow resources are not tied to a particular `SonataFlow` instance, just referenced by it. It means that the operator won't edit or delete them if the `SonataFlow` CR is updated or deleted. You have total control of the `ConfigMap` instance.

Reviewer 4: ```suggestion
The `ConfigMap` containing workflow resources is not tied to a particular `SonataFlow` instance, just referenced by it. It means that the operator won't edit or delete them if the `SonataFlow` CR is updated or deleted. You have total control of the `ConfigMap` instance.
```

> That said, you can organize your `ConfigMaps` in a way that other workflows could reuse them. In other words, one `ConfigMap` can be mapped to many `SonataFlow` Custom Resources.

Reviewer 4: ```suggestion
You can organize your `ConfigMaps` in a way that other workflows could reuse them. In other words, one `ConfigMap` can be mapped to many `SonataFlow` Custom Resources.
```

For example, if you have many OpenAPI or xref:eventing/orchestration-of-asyncapi-based-services.adoc[AsyncAPI] specification files that your workflows can access, you can group them by domain. This way you create a static Service Registry using `ConfigMaps`. Other developers within your company can reuse the same specification files when designing workflows.

The same applies for data input and output schemas, subflows definitions, and so on.

> 

Reviewer 5: Here is missing the types handled with Quarkus, 

- Camel Routes, 
- OpenAPI, 
- AsyncAPI, 
- Generic 

 and the default root path for workflowPath, `src/main/resources/`

Author: Doesn't matter the type anymore, actually. But I'll add the default mount path. 👍

Author: Actually, I don't think it matters for end users. This is an implementation detail, I won't add here.
